transformer_lens/HookedTransformer.py

```diff
@@ -271,126 +285,154 @@ def forward(
-            # See HookedTransformerConfig for details
-            pos_embed = self.hook_pos_embed(
-                self.pos_embed(tokens, pos_offset)
-            )  # [batch, pos, d_model]
-            residual = embed
-            shortformer_pos_embed = pos_embed
-        elif self.cfg.positional_embedding_type == "rotary":
-            # Rotary doesn't use positional embeddings, instead they're applied when dot producting keys and queries.
-            # See HookedTransformerConfig for details
-            residual = embed
-            shortformer_pos_embed = None
-        else:
-            raise ValueError(
-                f"Invalid positional_embedding_type passed in {self.cfg.positional_embedding_type}"
-            )
+        with utils.LocallyOverridenDefaults(
+            self, prepend_bos=prepend_bos, padding_side=padding_side
+        ):
+            if type(input) == str or type(input) == list:
+                # If text, convert to tokens (batch_size=1)
+                assert (
+                    self.tokenizer is not None
+                ), "Must provide a tokenizer if passing a string to the model"
+                # This is only intended to support passing in a single string
+                tokens = self.to_tokens(
+                    input, prepend_bos=prepend_bos, padding_side=padding_side
+                )
+            else:
+                tokens = input
+            if len(tokens.shape) == 1:
+                # If tokens are a rank 1 tensor, add a dummy batch dimension to avoid things breaking.
+                tokens = tokens[None]
+            if tokens.device.type != self.cfg.device:
+                tokens = tokens.to(devices.get_device_for_block_index(0, self.cfg))
+
+            if self.tokenizer and self.tokenizer.padding_side == "left":
+                # If the padding side is left, we need to compute the attention mask for the adjustment of
+                # absolute positional embeddings and attention masking so that the pad tokens are not attended.
+
+                if past_left_attention_mask is None:
+                    left_attention_mask = utils.get_attention_mask(
+                        self.tokenizer, tokens, self.cfg.default_prepend_bos
+                    )
+                else:
+                    assert (
+                        past_kv_cache is not None
+                    ), "If past_left_attention_mask is not None, past_kv_cache must not be None"
+                    assert (
+                        tokens.shape[1] == 1
+                    ), "If past_left_attention_mask is not None, tokens must be a single token along the sequence dimension"
+                    # past_kv_cache is not None, so we're doing caching.
+                    # We need to extend the past_left_attention_mask.
+                    # Append '1' to the right of the past_left_attention_mask to account for the new tokens
+                    left_attention_mask = utils.extend_tensor_with_ones(
```

Comment:
I couldn't think of a way that avoids introducing `past_left_attention_mask`. Any idea?